csharp/src/Google.Protobuf/MessageExtensions.cs

```diff
@@ -140,6 +143,53 @@ public static ByteString ToByteString(this IMessage message)
             return ByteString.AttachBytes(message.ToByteArray());
         }
 
+        /// <summary>
+        /// Checks if all required fields in a message have values set. For proto3 messages, this returns true
+        /// </summary>
+        public static bool IsInitialized(this IMessage message)
+        {
+            if (message.Descriptor.File.Proto.Syntax != "proto2")
+            {
+                return true;
+            }
+
+            return message.Descriptor
+                .Fields
+                .InDeclarationOrder()
+                .All(f =>
```

Comment:
It's slower than a normal implementation however I couldn't think of a non-breaking way to introduce a code-generated version. I don't believe it will be used too often, since the consumer will have to invoke it manually after parsing (the library doesn't on its own).

Theoretically yes. As far as I know, protobuf reflection should work on Unity so this should too. But I don't know exactly how I can test it on Unity so I don't know for sure.